### ISSUE TITLE
Allow global dprint config file

### ISSUE THREAD

**Issue by khrj** [dprint/dprint#355]
I often use the dprint vscode extension, and I'm editing a file that's not part of the current folder. It'd be great to support a global dprint config file, something along the lines of `~/.dprint/dprintrc.json`.

A command called `dprint init-with-global` would also be great, and it'd just copy the global config file to your current directory.

**Comment by disrupted** [comment]
I agree. This would be great to have and make it easier to manage consistent formatting across projects.

**Comment by dsherret** [comment]
This is something that could be added to the VS code extension. I've opened https://github.com/dprint/dprint-vscode/issues/13

For the CLI, it's possible to specify a config file via the `--config` flag and you can specify a URL there or a file path on the file system. I don't want to encourage having a global configuration file as people should define configuration files per project.

That said... if you are on linux and only working from the home folder then it is possible to define a configuration file at `~/.dprint.json` (hidden file config name is supported in dprint >= 0.13.0) assuming you are working from a descendant folder of your home directory and the CLI will pick it up if it can't find a config file in a closer directory (on windows, you could put it in the root folder of wherever you do development from).

> This would be great to have and make it easier to manage consistent formatting across projects.

To ensure this, you should create a dprint configuration file per project, but you can host a configuration file somewhere and specify that in an `"extends"` property. For example:

```ts
{
  "$schema": "https://dprint.dev/schemas/v0.json",
  "incremental": true,
  "extends": "https://yourdomain/path/to/config/file.v1.json",
  // todo: specify includes and excludes here
}
```

Or even...

```ts
{
  "$schema": "https://dprint.dev/schemas/v0.json",
  "incremental": true,
  "extends": [
    "https://yourdomain/path/to/config/file.v1.json",
    "https://yourdomain/path/to/config/other.v1.json"
  ],
  // todo: specify includes and excludes here
}
```

See https://dprint.dev/config/ for more details.

**Comment by dsherret** [comment]
Comment from @insomnimus:

> Please support a global config file.
>
> Running dprint init is fine, however it always initializes with the defaults
and I always end up configuring it.
>
> I know i can pass the config file from the command line but a default one is many times better
than specifying it through the command line each time.
>
> I also tried adding .dprint.json at the root of my development directory
but this doesn't work well because everything under the root directory gets formatted.
>
> Another way of accomplishing this would be if dprint supported extensions like cargo.
For example some dprint-custominit in $PATH, where I myself
could write the executable.

I think it's worth exploring this a bit more.

**Comment by dsherret** [comment]
What I'm thinking here is essentially:

1. If the ancestor directory tree HAS a dprint configuration file in it (local config):
   - It will resolve to using that config and work as-is.
1. If the ancestor directory tree does NOT have a dprint config file in it:
   1. It will print a warning that it's using the global config.
   1. The global configuration will be similar to a local configuration and contain a list of plugins along with their configuration.
      - I'm not sure that having an "includes" or "excludes" is useful or makes sense so probably at the start these would be ignored.
   1. If there are any plugins out of date, it will print a warning MAX once a week that their plugins are out of date and they can run `dprint config update` to update their plugins.
   
The global configuration would be stored in the user's local application data. Since this isn't always that easy to find, I think it would be beneficial to add a `dprint config edit` command that would launch an editor to edit the local or global configuration depending on what folder the user is in.

**Comment by edwardloveall** [comment]
Thanks for considering this 🙂 My primary use case is formatting my git commit messages which I write in markdown so having a global config is exactly what I need. I have a little feedback on your proposed idea above:

* 1: 👍 
* 2-i: I'm not sure a warning would be helpful, or at least not by default. Maybe in a verbose mode? I generally treat warnings as a bad practice I should consider fixing, but this sounds like a feature not a bad practice. Is that how you see it?
* 2-ii: I think `includes` and `excludes` could make sense because they support patterns and not just specific files. e.g. "I want to format all markdown documents wherever I am". Treating both global and local configs the same also has the advantage (I assume) of making the configuration code simpler.
* 2-iii: that makes sense as a warning to me 👍

**Comment by edwardloveall** [comment]
And FWIW, I found a couple tools that do something similar:

* `bottom`: [code](https://github.com/ClementTsang/bottom/blob/04d1218f659389d0d6ce6bb74a6a869d0fd76084/src/lib.rs#L196-L229), [docs](https://clementtsang.github.io/bottom/nightly/configuration/config-file/default-config/)
* `pijul`: [code](https://nest.pijul.com/pijul/pijul:main/SXEYMYF7P4RZM.BJPQC), [docs](https://pijul.org/manual/configuration.html)
* `lsd`: [code](https://github.com/Peltoche/lsd/blob/master/src/config_file.rs#L131-L151), [docs](https://github.com/Peltoche/lsd#configuration)
* `tokei`: [code](https://github.com/XAMPPRocky/tokei/blob/master/src/config.rs#L57-L135), [docs](https://github.com/XAMPPRocky/tokei#configuration)
* `procs`: [code](https://github.com/dalance/procs/blob/master/src/main.rs#L181-L210), [docs](https://github.com/dalance/procs#configuration)

Here's also an interesting spec that some of these tools use, if that's helpful: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

**Comment by ZaLiTHkA** [comment]
howdy.. 🖖🏼

I started down the path of reseaching `dprint` earlier this evening, and one of my biggest questions was quite simply: "what happens when there's no local config file?" and suddenly I was falling down a rabbit hole..

so, I see this idea has already been around for a while, but as it's still open I'm going to ~~brav~~naively asuming it's still open for discussion.. 🤔 

---

first a little disclaimer: I have _always_ preferred the idea of **un**opinionated configuration for tools like this. I firmly believe that certain things, such as codebase formatting, should be explicitly defined by the developer(s) responsible for a project. this is also not something that should change very often at all, even (especially?) across multiple projects under the same group.

if you get tired of setting the same config for every project, well.. then get lazy and script it. heck, create an alias with your preferred command and arguments and call `dprintify` **once** whenever you generate a new project (which is most likely also a one-liner, if you do that often enough).

still, I fully appreciate the convenience of the more opinionated tools as well (I use Prettier.. **a lot**..), but I'm also all too keenly aware of how limiting these systems can be. for example, Prettier and "explicit multi-line formatting", ultimately resulting in a third-party plugin called [`prettier-plugin-multiline-arrays`](https://github.com/electrovir/prettier-plugin-multiline-arrays) being created for those specific cases (and yes, I use that too.. **a lot**..).

---

with that out of the way, I fully agree with @dsherret here:

> This is something that could be added to the VS code extension. I've opened [dprint/dprint-vscode#13](https://github.com/dprint/dprint-vscode/issues/13)

adding "default" or "base" configuration logic to an IDE makes perfect sense. in my mind, one of the biggest benefits of any Integrated Development Enviromnent is the numerous layers of "configurability" they provide.. alas, I fear I might effectively be preaching to the choir on _that_ front.

but at the same time, I believe the CLI can do a little more of the legwork here, affording an opportunity to bring CLI and IDE usage more inline with each other.

so, building on the concept of a `--global` CLI tool argument...

> For the CLI, it's possible to specify a config file via the `--config` flag and you can specify a URL there or a file path on the file system (see [dprint.dev/cli](https://dprint.dev/cli/)). I don't want to encourage having a global configuration file as people should define configuration files per project so that someone else who goes to work on that project can know what the config is for that project.

...what if this was effectively just an alias for something akin to `--config ~/.dprint.json`?

in this case, I would say the following guidelines would be sufficient for providing _near_-zero configuration, without sacrificing per-project customisations:

* **with** local project configuration, pick that and don't check any other possible config locations
* **without** local project configuration or **with** `--global` set, try use the file at `~/.dprint.json`.
* **without** any of the above, dprint has no rules to reference and, well, simply does nothing.. or prints a warning (CI/CD env?).. or throws an error (dev env?)..

naturally, the `--global` and `--config <path>` arguments would have to be mutually exclusive. in which case I would favour the explicit `--config` argument over anything else. (or throw a usage warning.. or error.. etc)

---

there is one point here that I feel might be a little too limiting though, highlighted below:

> That said... if you are on linux and <ins>**only working from a descendant directory of the home folder**</ins> then it is possible to define a configuration file at `~/.dprint.json` (hidden file config name is supported in dprint >= 0.13.0) and the CLI will pick it up if it can't find a config file in a closer directory (on windows, you could put it in the root folder of wherever you do development from).

why does the working directory need to be a descendant of the user's home folder? sure, my `workspace` folder lives within my user folder, and I assume many other peoples' do as well.. but I have dev friends (yay) who put theirs on a completely separate partition to their user profile folder.

I would say, _first_ search up the tree from the current working directory, _then_ check the explicit user home folder path afterwards if need be.

at the end of the day, developers are lazy... if we don't need to do something, we don't. and if we need to do something repeatedly, we script it. 😎 not so?

**Comment by NukeManDan** [comment]
Really would _love_ this feaure, outside the use of VSCode - for example, I use https://docs.helix-editor.com/ and would like to call dprint from anywhere on file types I have in a global config and also (presumably required) global cache of the plugins set there.

IMHO the ideal would be to have a default global setting for the cached plugin - if it matched the local config - be used instead of re-installing the plugin in that local `.cache/...` location.

**Comment by utkarshgupta137** [comment]
My use case is that I want to use dprint for all my formatting. My work projects all have dprint configuration, but most of my other projects don't. I don't think there is an easy way in most editors to say to try dprint & fallback to prettier if it fails. So I want to use dprint everywhere. So I want a fallback to a global file, but as others have noted, it is not preferable to have it in `$HOME`.

> Instead of a dotfile in the user's home directory, please consider storing in the standard config directory of the OS. This crate documents them pretty well: https://github.com/dirs-dev/directories-rs#readme

Please don't use this crate as it uses a really obscure path on macOS. Most users on macOS prefer the XDG paths. If you want to be able to do it easily, might I suggest `etcetera` instead (disclosure: I maintain it). If the maintainers are interested, I can raise a PR.

**Comment by hemedani** [comment]
It should be nice if we place global config at `~/.config/.dprint.json` or  `~/.config/dprint/.dprint.json`  or  `~/.config/dprint.json` or  `~/.config/dprint/dprint.json` .
Putting the config in the home folder is somehow not good. And it may cause problems in some projects to find the root.

Also, please write this point in the [document](https://dprint.dev/config/).

**Comment by andreasputlitz** [comment]
Not sure, if this is a feature worth considering, but I wanted to add to the perspective.

---

For some context, what those "opinionated" formatters fundamentally get wrong in my opinion is the following: They want to solve a problem of "bikeshedding". This is implies two assumptions, which I think are wrong to begin with:

(1) Bikeshedding implies people fighting over questions of taste which in practice are virtually irrelevant. I think everybody using dprint is on the same page here, that code formatting is closely tied to code readability, which is closely tied to bugs, changebility, developer productivity / fatigue etc. - ergo _not_ merely an irrelevant question of taste. 

(2) But there is a second assumption here, which I think is not correct. Bikeshedding also implies that we are fighting over _the one correct final solution_, everybody has to adhere to. In this sense "opinionated" code formatter don't even "solve" that problem in a sense, because they move the one forced solution a higher up and affect even more people by it, which creates even steeper resistance. 

---

And I don't agree and even understand the necessity for this.

(a) First of all, I think most people would agree, that code readibility and developer productivity are at least in large parts individual - this is why we are here using a formatter with options, because your needs are not necessarily my needs. But I would add on top of that, that code readibility for me is even _situational_:

Am I working on my laptop in the train, or at home on my three screen setup? Am I currently skimming a new code base to understand how it works, or tinkering deep in the methods of some very large, complex class? Am I tired or well rested? You get the point. 

(b) Second of all, I don't even understand the technical necessity for this: With code formatters written in Rust, we can format even large files in acceptable to no time with one key stroke. So, I can without any issue format file A into style B in a matter of probably under a second, and then re-format in into style C just as quickly and easily.

---

Addtionally, it is worth noting, one of the main advantages of "opinionated" formatters that people in favour of them bring up, is that they make it easy to enforce a uniform code style across an entire organisation, and not have everybody use their private style, and have 500 lines change in a PR due to formatting. 

---

Which brings me to my suggestion, because I think we can have our cake here and eat it too:

You could have a "global" organization-wide established config, which is imported e.g. from some web ressource. This global config is the mandatory standard for any pushed code anywhere, and could even be automatically included in the CI-pipeline. Or it is a solo dev's personal default config across all projects.

But, in parallel to that, I might have one, two or maybe even three local, personal, situational configs, which might be assigned to a hotkey of the VS Code extension. One I would maybe call "compact", one "generous" - and they format the code on my situational needs, depending on current task, workplace, and personal condition.

---

I am actually using this already having two different config files: one for compact formatting for skimming (braces on the same, 2chars indent, very wide printwidth etc.) and one for generous formatting for zooming in and working in the code (allman braces, 4 chars indent, medium printwidth etc.).

**Comment by YowaiCoder** [comment]
I just want to use dprint as my default formatter and see prettier on my computer no more, I don't care if it is opinionated or not, I just need a handy formatter that works everywhere, and if there is a config in the project, use it, do what you must.

**Comment by innocenzi** [comment]
My workaround for now is to add `.dprint.json` to `~/.gitignore` and add that `.dprint.json` file in every project. It's a bit cumbersome though, and it'd be great if dprint supported a global configuration, either in the core or as an IDE extension feature.

**Comment by dsherret** [comment]
I think I have it in my head how this one should be implemented. I think probably this will just read from a specified location on the file system or otherwise uses a config in the binary. Then I'll pattern off what I did here with deno fmt when someone does `dprint fmt`: https://github.com/denoland/deno/pull/30623 -- Will try to get this one done by the month end.

**Comment by pikeas** [comment]
@dsherret Any updates on this feature?

**Comment by dsherret** [comment]
This is now released in 0.51.0, but the global configuration file doesn't work that well in vscode. I opened https://github.com/dprint/dprint-vscode/issues/118 to look into fixing this so please follow that issue if you use vscode.

For other editors, please open a new specific issue if you encounter them.

Release notes: https://github.com/dprint/dprint/releases/tag/0.51.0
Docs: https://dprint.dev/global-config/

I'll keep this issue open until it seems to be working ok.

**Comment by Randalphwa** [comment]
Looking for details about global config support.
